src/java.base/share/classes/jdk/internal/foreign/MemorySessionImpl.java

```diff
@@ -189,7 +287,12 @@ public boolean isAlive() {
      */
     @ForceInline
     public void checkValidStateRaw() {
-        if (owner != null && owner != Thread.currentThread()) {
+        if (owner == Thread.currentThread()) {
+            if (state < OPEN) {
```

Comment:
Why is the liveness check copied here as well? (as it is also done outside this if). This looks like a microoptimization - if so, some documentation should be added as to why the code looks like this.

Reply:
Also, here's there's no acquire fence before reading the state in the shared case. I assume the reasoning is that, for non closeable sessions, we can either see `0` (meaning we are trying to access a segment on that session before the session has been made non-closeable) or that we see `1` (meaning that the session has been fully constructed). So, in both cases it's ok to let the check go.

Reply:
That was my bad, the intention is to optimise for confined cases to reduce 1 comparison, just swapping the `&&` operands work as well, though.

Yes `acquireFence` is not needed as you said.